File: server/controllers/jobController.js
const { v4: uuidv4 } = require('uuid'); // Für eindeutige Job-IDs

// In-memory Job-Liste (für Demo-Zwecke)
// In einer echten Anwendung würdest du hier eine Datenbank verwenden
let jobs = [];

// Hilfsfunktion zur Simulation von Job-Verarbeitung
const processJob = (job) => {
    console.log(`Starting job ${job.id} (${job.type})...`);
    job.status = 'running';
    job.createdAt = new Date().toISOString();
    job.completedAt = null;
    job.error = null;
    job.progress = 0;

    const duration = Math.random() * 10000 + 2000; // 2-12 Sekunden
    const interval = setInterval(() => {
        job.progress = Math.min(job.progress + Math.random() * 10, 100);
        if (job.progress >= 100) {
            clearInterval(interval);
            const isSuccess = Math.random() > 0.2; // 80% Erfolg
            job.status = isSuccess ? 'completed' : 'failed';
            job.completedAt = new Date().toISOString();
            job.progress = isSuccess ? 100 : undefined; // Fortschritt bei Fehler zurücksetzen
            job.error = isSuccess ? null : `Simulierter Fehler bei der Verarbeitung.`;
            
            if (isSuccess) {
                 // Simulierte Artefakte erstellen
                 job.artifacts = [{
                    name: `output_${job.id.slice(0, 8)}.${job.type === 'convert' ? 'pdf' : job.type === 'tms' ? 'zip' : 'txt'}`,
                    type: job.type === 'convert' ? 'application/pdf' : job.type === 'tms' ? 'application/zip' : 'text/plain',
                    size: Math.floor(Math.random() * 5000000) + 100000, // 100KB - 5MB
                    url: `/api/jobs/${job.id}/artifacts/output`, // Beispiel-URL
                    viewable: job.type === 'convert' // Nur PDF ist "viewable" im Browser
                 }];
            } else {
                job.artifacts = [];
            }

            console.log(`Job ${job.id} finished with status: ${job.status}`);
        }
    }, duration / 10); // Update Fortschritt 10 mal

    // Speichere den Interval-ID, um ihn bei Abbruch zu löschen
    job._intervalId = interval;
};

// Initial einige Mock-Jobs erstellen, wenn das Backend startet
const generateInitialJobs = () => {
    const jobTypes = ['upload', 'convert', 'tms'];
    const initialJobs = [];
    for (let i = 0; i < 5; i++) {
        const type = jobTypes[Math.floor(Math.random() * jobTypes.length)];
        const id = uuidv4();
        const job = {
            id: id,
            name: `${type.charAt(0).toUpperCase() + type.slice(1)} Job ${i + 1}`,
            type: type,
            status: 'queued',
            createdAt: new Date().toISOString(),
            inputFile: {
                name: `initial_file_${i + 1}.${type === 'upload' ? 'dxf' : type === 'convert' ? 'dxf' : 'pdf'}`,
                size: Math.floor(Math.random() * 10000000) + 500000
            },
            parameters: type === 'convert'
                ? { pageSize: 'A4', dpi: 300 }
                : type === 'tms'
                ? { maxzoom: 18, format: 'png' }
                : {},
            artifacts: [],
            error: null,
            _intervalId: null // Zum Speichern des Simulations-Intervals
        };
        initialJobs.push(job);
    }
    jobs = initialJobs;
    // Starte die Verarbeitung für die initialen Jobs (simuliert)
    jobs.forEach(job => processJob(job));
};

// Beim ersten Laden des Controllers initiale Jobs generieren
generateInitialJobs();

// Controller-Funktionen

// Alle Jobs abrufen
exports.getJobs = (req, res) => {
    // Rückgabe einer Kopie, um direkte Modifikationen von außen zu verhindern
    res.json(jobs.map(({ _intervalId, ...job }) => job)); // _intervalId nicht zurückgeben
};

// Job abbrechen
exports.cancelJob = (req, res) => {
    const jobId = req.params.id;
    const job = jobs.find(j => j.id === jobId);

    if (!job) {
        return res.status(404).json({ message: 'Job not found' });
    }

    if (['running', 'processing', 'queued', 'pending'].includes(job.status)) {
        job.status = 'cancelled';
        job.completedAt = new Date().toISOString();
        job.error = 'Job wurde abgebrochen.';
        job.progress = undefined;
        if (job._intervalId) {
            clearInterval(job._intervalId);
        }
        console.log(`Job ${jobId} cancelled.`);
        // Rückgabe des aktualisierten Jobs (ohne _intervalId)
        const { _intervalId, ...updatedJob } = job;
        res.json(updatedJob);
    } else {
        res.status(400).json({ message: `Job cannot be cancelled in status: ${job.status}` });
    }
};

// Job wiederholen
exports.retryJob = (req, res) => {
    const jobId = req.params.id;
    const job = jobs.find(j => j.id === jobId);

    if (!job) {
        return res.status(404).json({ message: 'Job not found' });
    }

    if (['failed', 'error', 'cancelled'].includes(job.status)) {
        console.log(`Retrying job ${jobId}...`);
        // Setze Job-Status zurück und starte die Verarbeitung neu
        job.status = 'queued';
        job.completedAt = null;
        job.error = null;
        job.artifacts = [];
        job.progress = undefined; // Fortschritt zurücksetzen
        
        // Simulierte Verzögerung vor dem Neustart
        setTimeout(() => {
             processJob(job); // Starte die simulierte Verarbeitung neu
        }, 1000); // 1 Sekunde warten, bevor die Verarbeitung beginnt

        // Rückgabe des aktualisierten Jobs (ohne _intervalId)
        const { _intervalId, ...updatedJob } = job;
        res.json(updatedJob);

    } else {
        res.status(400).json({ message: `Job cannot be retried in status: ${job.status}` });
    }
};

// Job löschen
exports.deleteJob = (req, res) => {
    const jobId = req.params.id;
    const initialLength = jobs.length;
    
    // Stoppe die Simulation, falls der Job läuft
    const jobToDelete = jobs.find(j => j.id === jobId);
    if (jobToDelete && jobToDelete._intervalId) {
        clearInterval(jobToDelete._intervalId);
    }

    jobs = jobs.filter(j => j.id !== jobId);

    if (jobs.length < initialLength) {
        console.log(`Job ${jobId} deleted.`);
        res.status(204).send(); // 204 No Content bei erfolgreichem Löschen
    } else {
        res.status(404).json({ message: 'Job not found' });
    }
};

// Beispiel-Route für Artefakt-Download (simuliert)
// In einer echten Anwendung würdest du hier die Datei vom Speicherort ausliefern
router.get('/:id/artifacts/:artifactName', (req, res) => {
    const jobId = req.params.id;
    const artifactName = req.params.artifactName;
    const job = jobs.find(j => j.id === jobId);

    if (!job || !job.artifacts) {
        return res.status(404).json({ message: 'Job or artifacts not found' });
    }

    const artifact = job.artifacts.find(a => a.name.includes(artifactName)); // Einfache Namensprüfung

    if (!artifact) {
        return res.status(404).json({ message: 'Artifact not found' });
    }

    // Simulierte Dateiausgabe
    res.setHeader('Content-Type', artifact.type);
    res.setHeader('Content-Disposition', `attachment; filename="${artifact.name}"`);
    // Sende einfach einen Dummy-Text als Dateiinhalt
    res.send(`Simulierter Inhalt für Artefakt: ${artifact.name} von Job ${jobId}`);
});

// Füge die Artefakt-Route zum Router hinzu
router.get('/:id/artifacts/:artifactName', exports.downloadArtifact);

// Exportiere auch die Download-Funktion, falls sie direkt benötigt wird (optional)
exports.downloadArtifact = (req, res) => {
    // Diese Funktion wird jetzt von der Router-Definition oben verwendet
    // Der Code wurde in die Router-Definition verschoben, um den Kontext zu haben
    // Du könntest die Logik auch hier belassen und im Router aufrufen:
    // router.get('/:id/artifacts/:artifactName', (req, res) => exports.downloadArtifact(req, res));
    // Aber die direkte Implementierung im Router ist hier einfacher.
    // Belassen wir den Export für Konsistenz, auch wenn er im aktuellen Setup nicht direkt von außen aufgerufen wird.
};

File: server/controllers/jobController.py
import uuid
import datetime
import time
import random
import threading # Für die simulierte Job-Verarbeitung im Hintergrund
import logging

# Logging-Konfiguration
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# In-memory Job-Liste (für Demo-Zwecke)
# In einer echten Anwendung würdest du hier eine Datenbank verwenden
jobs_db = {} # Dictionary, um Jobs per ID zu speichern
job_processing_threads = {} # Um laufende Simulations-Threads zu verwalten {job_id: thread_object}
thread_lock = threading.Lock() # Lock für den Zugriff auf job_processing_threads

def process_job_simulation(job_id):
    """Simuliert die Verarbeitung eines Jobs."""
    job = jobs_db.get(job_id)
    if not job:
        return

    logging.info(f"Starting job {job['id']} ({job['type']})...")
    job['status'] = 'running'
    job['progress'] = 0
    
    # Simulierter Fehler nach einer gewissen Zeit für manche Jobs
    # fail_after_seconds = random.randint(5, 15) if random.random() < 0.3 else float('inf')
    # start_time = time.time()

    duration = random.uniform(2, 12)  # 2-12 Sekunden
    steps = 10
    step_duration = duration / steps

    for i in range(steps):
        with thread_lock:
            # Überprüfen, ob der Thread noch aktiv sein soll oder der Job abgebrochen wurde
            if job_id not in job_processing_threads or \
               threading.current_thread() != job_processing_threads.get(job_id) or \
               job.get('status') == 'cancelled':
                logging.info(f"Job {job_id} processing was stopped or cancelled.")
                # Wenn abgebrochen, setze den Status und die Endzeit, falls nicht schon geschehen
                if job.get('status') == 'cancelled' and not job.get('completedAt'):
                    job['completedAt'] = datetime.datetime.now().isoformat()
                    job['error'] = job.get('error', 'Job wurde während der Verarbeitung abgebrochen.')
                    job['progress'] = None
                # Entferne den Thread nur, wenn er der aktuelle ist und abgebrochen wurde
                if job_id in job_processing_threads and threading.current_thread() == job_processing_threads.get(job_id):
                    del job_processing_threads[job_id]
                return

        time.sleep(step_duration)
        # Stelle sicher, dass der Job noch existiert und nicht während des Sleeps gelöscht wurde
        if job_id not in jobs_db:
            logging.warning(f"Job {job_id} disappeared during processing sleep.")
            return
        job['progress'] = min(job.get('progress', 0) + (100 / steps), 100)

    # Job-Abschluss
    is_success = random.random() > 0.2  # 80% Erfolg
    job['status'] = 'completed' if is_success else 'failed'
    job['completedAt'] = datetime.datetime.now().isoformat()
    job['progress'] = 100 if is_success else None
    job['error'] = None if is_success else f"Simulierter Fehler bei der Verarbeitung von Job {job_id}."

    if is_success:
        job['artifacts'] = [{
            'name': f"output_{job['id'][:8]}.{job['type'] == 'convert' and 'pdf' or job['type'] == 'tms' and 'zip' or 'txt'}",
            'type': job['type'] == 'convert' and 'application/pdf' or job['type'] == 'tms' and 'application/zip' or 'text/plain',
            'size': random.randint(100000, 5000000), # 100KB - 5MB
            'url': f"/api/jobs/{job['id']}/artifacts/output", # Beispiel-URL
            'viewable': job['type'] == 'convert'
        }]
    else:
        job['artifacts'] = []
    
    logging.info(f"Job {job['id']} finished with status: {job['status']}")
    with thread_lock:
        if job_id in job_processing_threads and threading.current_thread() == job_processing_threads.get(job_id):
            del job_processing_threads[job_id]


def get_all_jobs():
    return list(jobs_db.values())

def create_initial_jobs():
    """Erstellt einige initiale Mock-Jobs beim Start."""
    # Diese Funktion sollte nur einmal ausgeführt werden, z.B. beim App-Start.
    if jobs_db or job_processing_threads: # Nur erstellen, wenn alles leer ist
        return

    job_types = ['upload', 'convert', 'tms']
    for i in range(5):
        job_type = random.choice(job_types)
        job_id = str(uuid.uuid4())
        created_at = datetime.datetime.now() - datetime.timedelta(minutes=random.randint(1, 60))
        
        job = {
            'id': job_id,
            'name': f"{job_type.capitalize()} Job {i + 1}",
            'type': job_type,
            'status': 'queued', # Startet als 'queued'
            'createdAt': created_at.isoformat(),
            'completedAt': None,
            'inputFile': {
                'name': f"initial_file_{i+1}.{job_type == 'upload' and 'dxf' or job_type == 'convert' and 'dxf' or 'pdf'}",
                'size': random.randint(500000, 10000000)
            },
            'parameters': {'pageSize': 'A4', 'dpi': 300} if job_type == 'convert' else \
                          {'maxzoom': 18, 'format': 'png'} if job_type == 'tms' else {},
            'artifacts': [],
            'error': None,
            'progress': None
        }
        jobs_db[job_id] = job
        # Starte die Simulation für diesen Job in einem neuen Thread
        with thread_lock:
            thread = threading.Thread(target=process_job_simulation, args=(job_id,))
            job_processing_threads[job_id] = thread
            thread.start()

def cancel_job_by_id(job_id):
    job = jobs_db.get(job_id)
    if not job:
        return None, "Job not found"

    if job['status'] in ['running', 'processing', 'queued', 'pending']:
        with thread_lock:
            job['status'] = 'cancelled'
            if not job.get('completedAt'): # Setze nur, wenn nicht schon gesetzt (z.B. durch Thread selbst)
                job['completedAt'] = datetime.datetime.now().isoformat()
            job['error'] = job.get('error', 'Job wurde abgebrochen.')
            job['progress'] = None
            
            # Der laufende Thread für diesen Job wird dies beim nächsten Check bemerken
            # und sich selbst beenden und aus job_processing_threads entfernen.
            # Wenn der Job 'queued' war und der Thread noch nicht gestartet ist,
            # wird er beim Start den 'cancelled' Status sehen.
            
        logging.info(f"Job {job_id} marked as cancelled.")
        return job, None # Rückgabe des Jobs, damit der Aufrufer den aktualisierten Status hat
    else:
        return None, f"Job cannot be cancelled in status: {job['status']}"

def retry_job_by_id(job_id):
    job = jobs_db.get(job_id)
    if not job:
        return None, "Job not found"

    if job['status'] in ['failed', 'error', 'cancelled']:
        logging.info(f"Retrying job {job_id}...")
        # Setze Job-Status zurück und starte die Verarbeitung neu
        job['status'] = 'queued'
        job['completedAt'] = None
        job['error'] = None
        job['artifacts'] = []
        job['progress'] = None
        job['createdAt'] = datetime.datetime.now().isoformat() # Update Erstellungszeit für Retry

        # Starte die Simulation für diesen Job in einem neuen Thread
        with thread_lock:
            # Stelle sicher, dass kein alter Thread für diesen Job mehr läuft oder registriert ist
            if job_id in job_processing_threads:
                old_thread = job_processing_threads.get(job_id)
                if old_thread and old_thread.is_alive():
                    logging.warning(f"Previous processing thread for job {job_id} was still alive during retry. This might lead to issues if not handled by the thread itself.")
                    # In einer robusten Implementierung müsste man hier den alten Thread sauber beenden.
                    # Für diese Simulation setzen wir darauf, dass der alte Thread den geänderten Job-Status erkennt.
                del job_processing_threads[job_id] # Entferne alte Referenz
            
            thread = threading.Thread(target=process_job_simulation, args=(job_id,))
            job_processing_threads[job_id] = thread
            thread.start()
        return job, None
    else:
        return None, f"Job cannot be retried in status: {job['status']}"

def delete_job_by_id(job_id):
    if job_id in jobs_db:
        with thread_lock:
            job_to_delete = jobs_db.get(job_id)
            if job_to_delete:
                # Markiere den Job als 'cancelled', um den laufenden Thread (falls vorhanden) zu informieren
                job_to_delete['status'] = 'cancelled' 
                # Der Thread wird sich selbst aus job_processing_threads entfernen.
                # Wenn der Thread nicht läuft, passiert nichts Schlimmes.

            # Entferne den Job aus der Datenbank
            if job_id in jobs_db:
                 del jobs_db[job_id]
            # Der Thread, falls er noch in job_processing_threads war, wird beim nächsten Check bemerken,
            # dass der Job weg ist oder 'cancelled' und sich beenden.
        logging.info(f"Job {job_id} deleted.")
        return True
    return False

def get_job_artifact(job_id, artifact_name_part):
    job = jobs_db.get(job_id)
    if not job or not job.get('artifacts'):
        return None, "Job or artifacts not found"

    for artifact in job['artifacts']:
        # Einfache Namensprüfung, da wir nur 'output' als Teil erwarten
        if artifact_name_part in artifact['name']:
            return artifact, None
    
    return None, "Artifact not found"

# create_initial_jobs() # Wird nun von app.py aufgerufen, um mehr Kontrolle zu haben
                        # oder kann hier bleiben, wenn das Modul nur einmal importiert wird.
                        # Für Flask mit `debug=True` (Auto-Reload) ist es besser, dies von app.py zu steuern.
